File: 1331-path-with-maximum-gold/path-with-maximum-gold.cpp
class Solution {
public:
    int n,m;
    int solve(vector<vector<int>>& grid,int i,int j)
    {
        if(i>=n or j>=m or i<0 or j<0 or grid[i][j]==0)
        return 0;

        int gold=grid[i][j];
        grid[i][j]=0;                 // marking the grid as visited.

        int up=solve(grid,i-1,j);
        int down=solve(grid,i+1,j);
        int left=solve(grid,i,j-1);
        int right=solve(grid,i,j+1);

        grid[i][j]=gold;

        // return grid[i][j] + std::max({up, down, right, left});
        return grid[i][j]+max(max(up,down),max(right,left));
    }

    int getMaximumGold(vector<vector<int>>& grid) 
    {
        int maxGold=0;
        n=grid.size();
        m=grid[0].size();
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(grid[i][j]!=0)
                {
                    maxGold=max(maxGold,solve(grid,i,j));
                }
            }
        }
        return maxGold;
    }
};

File: 1331-path-with-maximum-gold/path-with-maximum-gold.java
class Solution {
    int n, m; 
    int[][] directions ={{0,1},{0,-1},{1,0},{-1,0}};
    private int solve(int row,int col,int[][] grid)
    {
        if(row < 0 || row >= n || col < 0 || col >= m || grid[row][col] == 0)
        return 0;

        int gold = grid[row][col];
        grid[row][col] = 0;   // marking the cell as visited.

        int maxGold = 0; 
        for(int[] dir : directions)
        maxGold = Math.max(maxGold,solve(row + dir[0], col + dir[1], grid));

        grid[row][col] = gold; // unmarking the cell.
        return maxGold+gold;
    }

    public int getMaximumGold(int[][] grid) 
    {
        n = grid.length;
        m = grid[0].length;
        int maximumGold = 0;
        for(int i=0;i<n;i++)
        {
            for(int j = 0; j < m; j++)
            if(grid[i][j] > 0)
            maximumGold = Math.max(maximumGold, solve(i,j,grid));
        }
        return maximumGold;
    }
}